### PR TITLE
bugfix/28369_libreoffice_saga_continues: Introduce even more brutal paranoia in the LibreOffice scanner

### PR DESCRIPTION
**Pull request by magenta-af**
Yet _another_ unexpected LibreOffice failure mode popped up after the last release: `unoconv` processes can _also_ get stuck. When they do, they somehow remain connected to the UNIX pipe that the `soffice.bin` process communicates on: when we then start a new `soffice.bin` process, its messages get randomly shredded, and the fragments get divided amongst the new and old `unoconv` processes.

_Sigh._

This branch introduces `unoconv` lifecycle management, and also adds ridiculous paranoia around every attempt to interact with LibreOffice processes, signalling processes and conversion processes.

### CODE REVIEW

File: scrapy-webscanner/scanners/processors/libreoffice.py [hunk 1]
```diff
@@ -160,7 +161,12 @@ def convert(self, item, tmp_dir):
 
         attempts = 0
         while attempts < 4:
-            return_code = subprocess.call(unoconv_args)
+            return_code = 113
+            self.unoconv = subprocess.Popen(unoconv_args)
+            try:
+                return_code = self.unoconv.wait()
```
Comment (dals83) [inline scrapy-webscanner/scanners/processors/libreoffice.py]: What if self.unoconv.wait() gets stuck?

Comment (magenta-af) [reply scrapy-webscanner/scanners/processors/libreoffice.py]: Hm. That's a good point - this logic won't do what I wanted.

The idea here was that `self.unoconv` should be unset once `unoconv` exits successfully, to allow `teardown_queue_processing` to do its job, but if I always unset it in the `finally` block, then that won't work.

**Comment by magenta-af** [comment]
> something _simpler_

The LibreOffice backend, `soffice.bin`, can - and quite often does - freeze on large documents, and when it does this, the `unoconv` process communicating with it also freezes. It is possible to connect to an existing `soffice.bin` instance to ask it to cleanly shut down its communications channels, and we do try to do that, but the program that gets started in order to do that can itself freeze if the backend is frozen.

If we don't manage the backend ourselves, then `unoconv` will connect to a system-wide instance, starting it if necessary -- but a single instance is not actually capable of processing requests from several sources at once and so will often freeze. If we _do_ manage the backend ourselves, then we also need to manage the `unoconv` instances that talk to it: if the backend freezes and is restarted, but there's a stuck `unoconv` process in the middle of communicating with it, then that process will remain as a reader on the UNIX pipe, randomly consuming fragments of messages according to the whim of the scheduler, and so will make all future communications over that pipe _also_ (effectively) freeze.

The code must be able to recover as cleanly as possible from any combination of these situations, and I'm afraid I don't know of a simple way of doing that. Do you have any specific suggestions?

**Comment by danchr** [comment]
> Jeg ved ikke hvad alternativet til libreoffice skulle være. Men jeg kunne godt tænke mig at få lukket den her så jeg kan få den over på testserveren.

My gripe isn't with LibreOffice per sé, but with how we invoke it. It all just seems a bit iffy, but as Alec mentions in a comment the alternative is to “rewrite the processor to use pyUNO” — which sounds like a significant task. I can't see anything _obviously wrong_ in this PR, but then again, I don't feel comfortable saying that it's _good_ either. I hope that makes sense?

**Pull request merged**